File: packages/typescript/src/resources/prompts.ts
import type { HttpClient } from '../http.js';
import type {
  Prompt,
  PromptVersion,
  CreatePromptBody,
  UpdatePromptBody,
  CreatePromptVersionBody,
  PromptDiffResult,
} from '../types/prompt.js';

export class PromptsResource {
  constructor(private http: HttpClient) {}

  async list(): Promise<Prompt[]> {
    return this.http.get<Prompt[]>('/v1/prompts');
  }

  async get(id: string): Promise<Prompt & { latest?: PromptVersion }> {
    return this.http.get<Prompt & { latest?: PromptVersion }>(`/v1/prompts/${id}`);
  }

  async create(body: CreatePromptBody): Promise<Prompt> {
    return this.http.post<Prompt>('/v1/prompts', body);
  }

  async update(id: string, body: UpdatePromptBody): Promise<Prompt> {
    return this.http.patch<Prompt>(`/v1/prompts/${id}`, body);
  }

  async delete(id: string): Promise<{ ok: boolean }> {
    return this.http.delete<{ ok: boolean }>(`/v1/prompts/${id}`);
  }

  async listVersions(id: string): Promise<PromptVersion[]> {
    return this.http.get<PromptVersion[]>(`/v1/prompts/${id}/versions`);
  }

  async createVersion(id: string, body: CreatePromptVersionBody): Promise<PromptVersion> {
    return this.http.post<PromptVersion>(`/v1/prompts/${id}/versions`, body);
  }

  async getVersion(id: string, version: number): Promise<PromptVersion> {
    return this.http.get<PromptVersion>(`/v1/prompts/${id}/versions/${version}`);
  }

  async diff(fromId: string, toId: string): Promise<PromptDiffResult> {
    return this.http.get<PromptDiffResult>('/v1/prompts/diff', {
      params: { from: fromId, to: toId },
    });
  }
}

File: packages/python/invariance/resources/prompts.py
from __future__ import annotations

from typing import Any

from ..http_client import HttpClient


class PromptsResource:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    async def list(self) -> list[dict[str, Any]]:
        return await self._http.get("/v1/prompts")

    async def get(self, id: str) -> dict[str, Any]:
        return await self._http.get(f"/v1/prompts/{id}")

    async def create(self, body: dict[str, Any]) -> dict[str, Any]:
        return await self._http.post("/v1/prompts", body)

    async def update(self, id: str, body: dict[str, Any]) -> dict[str, Any]:
        return await self._http.patch(f"/v1/prompts/{id}", body)

    async def delete(self, id: str) -> dict[str, Any]:
        return await self._http.delete(f"/v1/prompts/{id}")

    async def list_versions(self, id: str) -> list[dict[str, Any]]:
        return await self._http.get(f"/v1/prompts/{id}/versions")

    async def create_version(
        self, id: str, body: dict[str, Any]
    ) -> dict[str, Any]:
        return await self._http.post(f"/v1/prompts/{id}/versions", body)

    async def get_version(self, id: str, version: int) -> dict[str, Any]:
        return await self._http.get(f"/v1/prompts/{id}/versions/{version}")

    async def diff(self, from_id: str, to_id: str) -> dict[str, Any]:
        return await self._http.get(
            "/v1/prompts/diff", params={"from": from_id, "to": to_id}
        )
